src/Microsoft.AspNet.Mvc.Core/Filters/FormatFilterAttribute.cs

```diff
@@ -0,0 +1,133 @@
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Class | AttributeTargets.Method, AllowMultiple = false, Inherited = true)]
+    public class FormatFilterAttribute : Attribute, IFormatFilter, IResourceFilter, IResultFilter
+    {
+        /// <summary>
+        /// As a <see cref="IResourceFilter"/>, this filter looks at the request and rejects it before going ahead if
+        /// 1. The format in the request doesnt match any format in the map.
+        /// 2. If there is a conflicting producesFilter.
+        /// </summary>
+        /// <param name="context">The <see cref="ResourceExecutingContext"/>.</param>
+        public void OnResourceExecuting([NotNull] ResourceExecutingContext context)
+        {
+            var format = GetFormat(context);
+
+            if (!string.IsNullOrEmpty(format))
+            {
+                var formatContentType = GetContentType(format, context);
+                if (formatContentType == null)
+                {
+                    // no contentType exists for the format, return 404
+                    context.Result = new HttpNotFoundResult();
+                }
+                else
+                {
+                    var responseTypeFilters = context.Filters.OfType<IApiResponseMetadataProvider>();                    
+                    var contentTypes = new List<MediaTypeHeaderValue>();
+
+                    foreach (var filter in responseTypeFilters)
+                    {
+                        filter.SetContentTypes(contentTypes);
+                    }
+
+                    if (contentTypes.Count != 0)
+                    {
+                        // If formatfilterContentType is not subset of any of the content types produced by 
+                        // IApiResponseMetadataProviders, return 404
+                        if (!contentTypes.Any(c => formatContentType.IsSubsetOf(c)))
+                        {
+                            context.Result = new HttpNotFoundResult();
+                        }
+                    }                   
+                }
+            }
+        }
+
+        /// <inheritdoc />
+        public void OnResourceExecuted([NotNull] ResourceExecutedContext context)
+        {
+
+        }
+
+        /// <summary>
+        /// Sets a Content Type on an  <see cref="ObjectResult" />  using a format value from the request.
+        /// </summary>
+        /// <param name="context">The <see cref="ResultExecutingContext"/>.</param>
+        public void OnResultExecuting([NotNull] ResultExecutingContext context)
+        {
+            var format = GetFormat(context);
+            if (!string.IsNullOrEmpty(format))
+            {
+                var objectResult = context.Result as ObjectResult;
+                if (objectResult != null)
+                {
+                    var contentType = GetContentType(format, context);
+                    Debug.Assert(contentType != null);
+                    objectResult.ContentTypes.Clear();
```

Comment:
space out the `debug.assert` from the functional code, and add a clear message to what went wrong. In this case something like: ContentType was provided in the resource stage, but now it's gone...

This is another good reason to move this to a filter factory, create the filter instance as `Transient` and store the ContentType at the resource filter step on a property of the filter.